### ISSUE TITLE
How to bypass DDR and directly stress EMMC?

### ISSUE THREAD

**Issue by MikeWu123** [stressapptest/stressapptest#50]
Can parameter -f bypass DDR and directly access EMMC?
I would like to separate test for these 2 components. Is it possible?

**Comment by nickjsanders** [comment]
Unfortunately, not really. EMMC is typically only accessible by DMA, which
stages data in RAM then transfers to EMMC.

**Comment by MikeWu123** [comment]
Thank for clarification, Nick.
If so, how do I distinguish if it's EMMC issue or DDR issue? For example, I met two case:
[Case 1]
Report Error: miscompare : DIMM Unknown : 1 : 39s
Hardware Error: miscompare on CPU 1(0x2) at 0x9efde0a0(0xa6c80a3:DIMM Unknown): read:0xffdfffffbfbfffff, reread:0xffdfffffbfbfffff expected:0xffdfffffffbfffff
[Case 2]
Report Error: write-error : hdUnknown : 1 : 6s 
Block Error: file_thread failed to write, bailing 

Another question, from https://github.com/stressapptest/stressapptest/blob/1d848e96ae0894bec29e6c5a503006c42c5719a0/src/os.cc#L628
void *map_buf = mmap(NULL, length, PROT_READ | PROT_WRITE,
MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
Your source code claim mmap with MAP_PRIVATE which is based on COPY on WRITE. Why don't you use MAP_SHARED which is write through to file itself?

**Comment by nickjsanders** [comment]
On Tue, Oct 24, 2017 at 5:21 PM, MikeWu123 <notifications@github.com> wrote:

> Thank for clarification, Nick.
> If so, how do I distinguish if it's EMMC issue or DDR issue? For example,
> I met two case:
> [Case 1]
> Report Error: miscompare : DIMM Unknown : 1 : 39s
> Hardware Error: miscompare on CPU 1(0x2) at 0x9efde0a0(0xa6c80a3:DIMM
> Unknown): read:0xffdfffffbfbfffff, reread:0xffdfffffbfbfffff
> expected:0xffdfffffffbfffff
>
Case 1 is probably a DRAM error, and definitely not an EMMC error. The data
was written to DRAM, and then read back, and had changed. It's also
possible but unlikely that it was corrupted in the CPU.


> [Case 2]
> Report Error: write-error : hdUnknown : 1 : 6s
> Block Error: file_thread failed to write, bailing
>
Case 2 is probably an EMMC error. The data was written to memory, then
written to the block device (EMMC), then read back into memory, and had
changed. This is likely caused by EMMC problems, although it's possible
that it was corrupted in DRAM or CPU, since the data passed through those
as well.



> Another question, from https://github.com/stressappte
> st/stressapptest/blob/1d848e96ae0894bec29e6c5a503006c42c5719
> a0/src/os.cc#L628
> void *map_buf = mmap(NULL, length, PROT_READ | PROT_WRITE,
> MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
> Your source code claim mmap with MAP_PRIVATE which is based on COPY on
> WRITE. Why don't you use MAP_SHARED which is write through to file itself?
>
This mmap is just a memory mapping from the shared memory allocation to the
process. It's not backed by anything on EMMC, so there's no file involved.
The issue is that 32bit ARM devices often have more DRAM than can be
addressed directly by a single linux process.

**Comment by MikeWu123** [comment]
In case 2, that's what I am confused. Even it says failure on block device, but can't judge it's DDR or EMMC problem. Do you have suggestions on testing EMMC purely?

**Comment by nickjsanders** [comment]
You can't test EMMC purely through software, since software runs in DRAM.

What I typically do is run the DRAM-only stressapptest to verify DRAM is
PASS, then run the EMMC test, and it's generally safe to assume that if
stressapptest fails now it's because of EMMC failures.

**Comment by MikeWu123** [comment]
Agree with you. Thank for your suggestion, Nick.

**Comment by chienminglo** [comment]
I've tried below 2 commands to test the emmc on my device:
./stressapptest -f ./file1 -f ./file2 -v 20 -s 30
    Stats: Completed: 168892.00M in 31.36s 5384.91MB/s, with 0 hardware incidents, 0 errors
    Stats: Memory Copy: 165692.00M at 5449.47MB/s
    Stats: File Copy: 3200.00M at 104.85MB/s
./stressapptest -f ./file1 -f ./file2 -m 0 -v 20 -s 30
    Stats: Completed: 5696.00M in 30.15s 188.94MB/s, with 0 hardware incidents, 0 errors
    Stats: Memory Copy: 0.00M at 0.00MB/s
    Stats: File Copy: 5696.00M at 188.96MB/s
If I set zero thread for memory copy, from the result it seems emmc is tested in more stressful way (i/o throughput increases obviously).
May I ask if in this way can run correctly? the result convincible? or any side effect will occur?

**Comment by nickjsanders** [comment]
Yes, by turning off the memory copy threads, the system load will be lower.
If there was previously some
resource contention between memory copy and the emmc driver, after
disabling memory copy, the emmc
performance would be higher.

I'd think your conclusion makes sense and the test you are running is
reasonable. The only side effect is that
disabling memory copy will remove test coverage for DRAM, obviously. But
this is a good thing if you are only
trying to test emmc.
